woocommerce/woocommerce-ios
# Product Subscriptions: Enable creating and editing variable subscription products

Part of #11171 


## What

This PR updates the product form for variable subscription products to enable generating/updating variations. This also updates the use case for generating variations to include subscriptions if the product is of type variable subscription.

## How
- Networking:
  - Updated `CreateProductVariation` with a new optional property for subscription. This will be encoded similarly to products and variations.
  - Updated unit tests.
- UI:
  - Updated `ProductFormActionsFactory` for variable subscription products to display similar fields as regular variable products.
  - Updated `ProductVariationFormActionsFactory` for variations of variable subscription products to display similar fields as regular variable products. We will need to add more subscription-related fields such as free trial and expiration date, which should be handled in a separate PR.
  - Updated `ProductFormRemoteActionUseCase` to support duplicating subscriptions for variations if available.
  - Updated `GenerateVariationUseCase` to create empty subscriptions when generating variations for variable subscription products.
  - Updated `ProductVariationGenerator` to create empty subscriptions when generating all variations for variable subscription products.
  - Updated `ProductFormViewController` to send the current product model instead of the original one to the variation controller. This is to fix the issue generating variation when changing from a different type to variable subscription.
  - Updated `ProductVariationsViewModel` to enable generating and bulk updating variations for variable subscription products when the feature flag is enabled.
  - Update tests.

## Testing instructions

Create a variable subscription product
- Enable the feature flag `subscriptionProducts` and build the app.
- On the app, log in to a store with Woo Subscription plugin set up.
- Navigate to Products tab > select "+" > select manual option > Variable subscription product.
- Notice that the product form displays similar fields as regular variable product's.
- Create at least 1 attribute for the product and generate either one or all variations.
- When done, open the variation list and select any item.
- Notice that the variation form displays similar fields as regular variations.
- Tap the Price field, notice that there are subscription-related fields: billing interval and sign-up fee.
- Feel free to try creating a regular variable product, the price settings screen in each variation should not display subscription-related fields.

Duplicate a variable subscription product.
- Follow the steps above to create a variable subscription product or create one on your store's wp-admin.
- Navigate to the Products tab and select the created variable subscription product.
- Tap the ellipsis menu "..." and select Duplicate.
- When done, the duplicated product should have the same variations with the original, and you should be able to find the subscription-related fields in the price settings screen of each variation.

## Screenshots


https://github.com/woocommerce/woocommerce-ios/assets/5533851/714b5739-49f6-4020-86a6-5fca2aa80447



---
